### PR TITLE
Build `nix-util` with Meson

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

The idea is two-fold:

- Replace autotools with Meson

- Build each library in its own derivation

The interaction of these two features is that Meson's "subprojects" feature (https://mesonbuild.com/Subprojects) allows us to have single dev shell for building all libraries still, while also building things separately. This allows us to break up the build without a huge productivity lost.

~~By doing the first two libraries, I hope to demonstrate how concerns are divided across the package divided, how configure checks (mostly) don't need to be duplicated, etc. Hopefully this this makes for an example that is easy enough to imitate so others can help out with this.~~ That is in `meson-libstore`. 

I tested the Linux native build, and NetBSD and Windows cross builds.

Also do some clean ups of the Flake in the process of supporting new jobs.

# Context

Special thanks to everyone that has worked on a Meson port so far, @p01arst0rm and @Qyriad in particular.

Depends on #10839 

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: .github/labeler.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 "c api":
   - changed-files:
-    - any-glob-to-any-file: "src/lib*-c/**/*"
+    - any-glob-to-any-file: "subprojects/lib*-c/**/*"
     - any-glob-to-any-file: "test/unit/**/nix_api_*"
```
Comment (L-as) [inline .github/labeler.yml:4]: Is this rename really needed?

Comment (L-as) [reply .github/labeler.yml:4]: Oh I see it's discussed in https://github.com/NixOS/nix/pull/10839#issuecomment-2145830083

Comment (Ericson2314) [reply .github/labeler.yml:4]: Turns out it is not!

**Comment by fricklerhandwerk** [comment]
This looks a actually tractable. The only question I have is whether there's a way around specifying every single source file. This sounds like guaranteed busywork.

**Comment by Ericson2314** [comment]
> The only question I have is whether there's a way around specifying every single source file. This sounds like guaranteed busywork.

Nope, sorry https://github.com/mesonbuild/meson/blob/master/docs/markdown/FAQ.md#why-cant-i-specify-target-files-with-a-wildcard

Someday dynamic derivations Nix will be a backend instead of Ninja for Meson too, and the underlying issues thus overcome, and then we can use wildcards again :).

**Review by fricklerhandwerk — APPROVED** [review]
In that case...

**Comment by L-as** [comment]
How will this interact with `make compile_commands.json`?

**Comment by Ericson2314** [comment]
https://github.com/mesonbuild/meson/pull/4547/files/52071c6d4ee15c750f8a8620e038b78627db890e#r381455882 I am a bit confused about the status of the docs (@eli-schwartz?) but Meson has a bunch of stuff out of the box for this

https://mesonbuild.com/IDE-integration.html

**Comment by eli-schwartz** [comment]
Yes, it might be a good idea to re-add mention of compile_commands.json to the docs.

Note that meson will always generate one at the end of `meson setup` without your intervention, on the theory that creating one is never harmful so why make people specify they want it.

**Comment by Ericson2314** [comment]
I'm happy to say that in the big dev shell, one can build the perl bindings too! No more forgetting to update them during development!

**Review by inclyc — APPROVED** [review]
So there are actually an option to use subprojects in different dirname. That's much better than previous PR.

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# This is just a stub project to include all the others as subprojects
+# for development shell purposes
+
+project('nix-dev-shell', 'cpp',
+  version : run_command('cat', './.version', check : true).stdout().strip(),
```
Comment (eli-schwartz) [inline meson.build:5]: What's the minimum version of meson you support? This doesn't list `meson_version: '>= 1.1'` -- you're using `meson.options` which requires this.

The perl subproject requires 0.64.0, and the earliest version of meson in repology for nixos is 0.57.1

Meson 0.57.0 added support for this:

```suggestion
  version : files('.version'),
```

upon which meson itself will read this file for you without forking out to `cat` (which in theory may not be in $PATH, and e.g. on Windows may not exist at all absent msys2 or cygwin).

File: src/libstore/meson.build [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+
+deps_private = [ ]
+deps_public = [ ]
+deps_other = [ ]
+
+configdata = {
+  # TODO rename, because it will conflict with downstream projects
+  'PACKAGE_VERSION' : '"' + meson.project_version() + '"',
+  'SYSTEM': '"' + host_machine.system() + '"',
+}
```
Comment (eli-schwartz) [inline src/libstore/meson.build:23]: Probably, configdata should be an instance of `configuration_data()` so that you can use `.set_quoted()` instead of implementing this via `'"' + var + '"'` which is not precisely the height of readability.

File: src/libutil/meson.build [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+  # avoid absolute paths, use vendored ones
+  libraries_private : libraries_private,
+)
+
+nix_util = declare_dependency(
+  include_directories : include_dirs,
+  link_with : libutil,
+  compile_args : ['-std=c++2a'],
+  dependencies : [],
+)
```
Comment (eli-schwartz) [inline src/libutil/meson.build:298]: Also add:

```
meson.override_dependency('nix-util', nix_util)
```

File: src/libstore/meson.build [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+
+nix_util = dependency('nix-util', required : false)
+nix_util_from_subproject = not nix_util.found()
+if nix_util_from_subproject
+  nix_util = subproject('libutil').get_variable('nix_util')
+  # subproject sadly no good for pkg-config module
```
Comment (eli-schwartz) [inline src/libstore/meson.build:29]: ... which then means you can simply do:

```meson
nix_util = dependency('nix-util')
nix_util_from_subproject = nix_util.type_name() == 'internal'
```
and since libutil is processed before libstore, it will be already overridden by the time you look it up here, and it will just use the subproject version.

Also re: knowing if it's from a subproject, I thought that since nix-util has been processed via pkgconfig.generate() before being processed by declare_dependency, meson is supposed to trace that and reuse that information via the pkgconfig module inside libstore. I could be misremembering, but I thought it was supposed to work.

File: src/libstore/meson.build [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+  }
+endif
+
+generated_headers = []
+foreach header : [ 'schema.sql', 'ca-specific-schema.sql' ]
+  generated_headers += custom_target(
+    command : [ 'bash', '-c', 'echo \'R"__NIX_STR(\' | cat - @INPUT@ && echo \')__NIX_STR"\'' ],
+    input : header,
+    output : '@PLAINNAME@.gen.hh',
+    capture : true,
```
Comment (eli-schwartz) [inline src/libstore/meson.build:134]: If you are already running a bash command, why run `capture: true`? This just means that meson will wrap the entire bash command inside a python wrapper script that is in charge of passing inputs/outputs.

What you really want is pretty simple:

```meson
    command: ['bash', '-c', 'echo \'R"__NIX_STR(\' | { cat - @INPUT@ && echo \')__NIX_STR"\'; } > "$1"', '_ignored_argv0', '@OUTPUT@'],

**Review by eli-schwartz — CHANGES_REQUESTED** [review]
Note that in various places, `required: true` is used but is also the default. Your choice whether "explicit is better than implicit" for default values, but it would make lines shorter to omit them, and possibly avoid having to break across multiple lines to play nice with line length style requirements.

File: src/libutil/meson.build (hunk 4 above)
Comment (eli-schwartz) [reply src/libutil/meson.build:298]: Aside for the benefits I mentioned below, this also means that you can't have two versions of nix-util pulled into your build, if there's a system copy and libstore picks that up but you are *also* building the libutil subproject as part of the same build.

**Comment by Ericson2314** [comment]
Thanks @eli-schwartz, those suggestions made it a lot better!

File: src/libstore/meson.build (hunk 5 above)
Comment (eli-schwartz) [reply src/libstore/meson.build]: I assume though that it is in fact required. The previous `required: false` was because you wanted to manually code the fallback.

File: meson.build (hunk 2 above)
Comment (eli-schwartz) [reply meson.build]: Note there are a couple other files that also use the `cat` approach, I just only reviewed the first location. :)

**Comment by fricklerhandwerk** [comment]
Tracking https://github.com/NixOS/nix/pull/10855

**Comment by edolstra** [comment]
I'm not saying I'm against switching to Meson, but I'm not convinced that the approach in this PR is the way to go. It adds a significant amount of complexity to our build system: src/libstore/meson.build is 446 lines (compared to 104 lines for src/libstore/local.mk, and unlike the makefile it's very imperative), src/libstore/package.nix is 158 lines, src/libutil/meson.build is 287 lines, etc. It doesn't make me feel confident that switching to Meson will make Nix easier to maintain.

There is probably a use case for having libstore exposed as a separate derivation (though for most users, having a `lib` output should be enough), but libutil is strictly internal to the Nix project and is not intended as an external artifact.

File: src/libstore/meson.build [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+  # at ~1% overhead in `nix search`.
+  #
+  # FIXME: remove when we get meson 1.4.0 which will default this to on for us:
+  # https://mesonbuild.com/Release-notes-for-1-4-0.html#ndebug-setting-now-controls-c-stdlib-assertions
+  '-D_GLIBCXX_ASSERTIONS=1',
+  language : 'cpp',
+)
+
+sources = files(
+  'binary-cache-store.cc',
```
Comment (edolstra) [inline src/libstore/meson.build:171]: I much prefer the current approach where we don't explicitly enumerate every source file, i.e.
```
nix_SOURCES := \
  $(wildcard $(d)/*.cc) \
```
A list like this causes a lot of unnecessary cod churn and tends to be a source of merge conflicts.

Is that possible with Meson?

File: src/libstore/meson.build [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+foreach funcspec : check_funcs
+  define_name = 'HAVE_' + funcspec.underscorify().to_upper()
+  define_value = cxx.has_function(funcspec).to_int()
+  configdata.set(define_name, define_value)
+endforeach
+
+has_acl_support = cxx.has_header('sys/xattr.h') \
+  and cxx.has_function('llistxattr') \
+  and cxx.has_function('lremovexattr')
+configdata.set('HAVE_ACL_SUPPORT', has_acl_support.to_int())
```
Comment (edolstra) [inline src/libstore/meson.build:71]: I wouldn't mind getting rid of Autoconf-style checks like this and just assuming that if you're on (for example) Linux, then you must have ACL support. So we can just do
```
#if __linux__
#include <sys/xattr.h>
#endif
```

File: src/libstore/meson.build (hunk 7 above)
Comment (p01arst0rm) [reply src/libstore/meson.build:171]: no, this is not possible by design. everything in meson is defined explicitly

File: src/libstore/meson.build (hunk 8 above)
Comment (p01arst0rm) [reply src/libstore/meson.build:71]: def a good idea, though the point of this PR is to change out the buildsystem pending future improvements

**Comment by p01arst0rm** [comment]
> I'm not saying I'm against switching to Meson, but I'm not convinced that the approach in this PR is the way to go. It adds a significant amount of complexity to our build system: src/libstore/meson.build is 446 lines (compared to 104 lines for src/libstore/local.mk, and unlike the makefile it's very imperative), src/libstore/package.nix is 158 lines, src/libutil/meson.build is 287 lines, etc. It doesn't make me feel confident that switching to Meson will make Nix easier to maintain.
> 
> There is probably a use case for having libstore exposed as a separate derivation (though for most users, having a `lib` output should be enough), but libutil is strictly internal to the Nix project and is not intended as an external artifact.

iirc the reason this was the way we chose is because having seperate derivations helps improve testing. It also helps to keep the source files for each library in its own place; currrently docgen is a mess, and de-mangling it involves moving the gen files into the /proj/lib directory. seperating them out this way massively improves maintainability into the future :)

File: src/libstore/meson.build (hunk 7 above)
Comment (eli-schwartz) [reply src/libstore/meson.build:171]: Meson has a FAQ entry on why this isn't natively supported:

https://mesonbuild.com/FAQ.html#why-cant-i-specify-target-files-with-a-wildcard

And another entry on how you can implement it yourself: 

https://mesonbuild.com/FAQ.html#but-i-really-want-to-use-wildcards

Comment (p01arst0rm) [reply src/libstore/meson.build:171]: code churn is a fair price to pay for a more robust build :)

File: src/libstore/package.nix [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+# Whether to use garbage collection for the Nix language evaluator.
+#
+# If it is disabled, we just leak memory, but this is not as bad as it
+# sounds so long as evaluation just takes places within short-lived
+# processes. (When the process exits, the memory is reclaimed; it is
+# only leaked *within* the process.)
+#
+# Temporarily disabled on Windows because the `GC_throw_bad_alloc`
+# symbol is missing during linking.
+, enableGC ? !stdenv.hostPlatform.isWindows
```
Comment (roberth) [inline src/libstore/package.nix:46]:
```suggestion
```

`libstore` does not use memory garbage collection.

File: src/libstore/package.nix [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+, sqlite
+
+, busybox-sandbox-shell ? null
+
+# Configuration Options
+#:
+# This probably seems like too many degrees of freedom, but it
+# faithfully reflects how the underlying configure + make build system
+# work. The top-level flake.nix will choose useful combinations of these
+# options to CI.
```
Comment (roberth) [inline src/libstore/package.nix:26]:
```suggestion
```

No need to apologize.

File: src/libstore/package.nix [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+, pname ? "nix-store"
+
+, versionSuffix ? ""
+, officialRelease ? false
+
+# Check test coverage of Nix. Probably want to use with with at least
+# one of `doCHeck` or `doInstallCheck` enabled.
```
Comment (roberth) [inline src/libstore/package.nix:34]:
```suggestion
# one of `doCheck` or `doInstallCheck` enabled.
```

File: src/libstore/package.nix [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+  postInstall =
+    # Remove absolute path to boost libs
+    ''
+      sed -i "$out/lib/pkgconfig/nix-store.pc" -e 's, ${lib.getLib boost}[^ ]*,,g'
+    '';
+
+  separateDebugInfo = !stdenv.hostPlatform.isStatic;
+
+  # TODO `releaseTools.coverageAnalysis` in Nixpkgs needs to be updated
+  # to work with `strictDeps`.
```
Comment (roberth) [inline src/libstore/package.nix:145]: (pre-existing, non-blocker)

What's going on here? Can we have an upstream issue link?

File: src/libstore/package.nix [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+      ./windows/meson.build
+      (fileset.fileFilter (file: file.hasExt "cc") ./.)
+      (fileset.fileFilter (file: file.hasExt "hh") ./.)
+      (fileset.fileFilter (file: file.hasExt "sb") ./.)
+      (fileset.fileFilter (file: file.hasExt "md") ./.)
+      (fileset.fileFilter (file: file.hasExt "sql") ./.)
+    ];
+  };
+
+  outputs = [ "out" "dev" ];
```
Comment (roberth) [inline src/libstore/package.nix:84]: We may consider pulling commonalities for the Nix libraries into a separate file, and use `lib.extends` to merge the functions overlay-style.

File: src/libutil/package.nix [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+, libsodium
+, nlohmann_json
+, openssl
+
+# Configuration Options
+#:
+# This probably seems like too many degrees of freedom, but it
+# faithfully reflects how the underlying configure + make build system
+# work. The top-level flake.nix will choose useful combinations of these
+# options to CI.
```
Comment (roberth) [inline src/libutil/package.nix:23]:
```suggestion
```

same

File: flake.nix [hunk 15]
```diff
@@ -255,23 +281,37 @@
...
+            name = "${pkgName}-static";
+            value = nixpkgsFor.${system}.static.${pkgName};
+          }))
+        // builtins.listToAttrs (lib.concatMap
+          (crossSystem: forAllPackagesList
+            (pkgName: {
+              name = "${pkgName}-${crossSystem}";
+              value = nixpkgsFor.${system}.cross.${crossSystem}.${pkgName};
+            }))
           crossSystems)
```
Comment (roberth) [inline flake.nix:295]: Let's simplify this to one Cartesian product on attrsets:

 - delete `forAllPackagesList`
 - GitHub won't let me make a suggestion that includes the 7 lines that come after this, but those can be deleted too.
 - this suggestion:
 
```suggestion
      } // lib.concatMapAttrs
        (pkgName: { }: {
          "${pkgName}-static" = nixpkgsFor.${system}.static.${pkgName};
          "${pkgName}-${crossSystem}" = nixpkgsFor.${system}.cross.${crossSystem}.${pkgName};
          "${pkgName}-${stdenvName}" = nixpkgsFor.${system}.stdenvs."${stdenvName}Packages".${pkgName};
        })
        { "nix" = { };
          "nix-store" = { };
          "nix-util" = { };
        })
```

File: flake.nix [hunk 16]
```diff
@@ -255,23 +281,37 @@
...
-            value = nixpkgsFor.${system}.stdenvs."${stdenvName}Packages".nix;
-          })
-          stdenvs)));
+        // builtins.listToAttrs (lib.concatMap
+          (stdenvName: forAllPackagesList
+            (pkgName: {
+              name = "${pkgName}-${stdenvName}";
+              value = nixpkgsFor.${system}.stdenvs."${stdenvName}Packages".${pkgName};
+            }))
+          stdenvs));
```
Comment (roberth) [inline flake.nix:302]:
```suggestion
```

See previous suggestion.

File: meson.build [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+  subproject_dir : 'src',
```
[the remark this answers is not in the document]
Comment (roberth) [reply meson.build:6]: I don't see any other mention of `nix-dev-shell`. How do we use this in practice?
Maybe add a paragraph to the hacking guide?

File: flake.nix (hunk 15 above)
Comment (roberth) [reply flake.nix:295]: Lists are no good, especially when all you need are attributes.

File: flake.nix [hunk 18]
```diff
@@ -288,12 +328,25 @@
...
+            BOOST_LIBRARYDIR = "${lib.getLib pkgs.boost}/lib";
             # For `make format`, to work without installing pre-commit
             _NIX_PRE_COMMIT_HOOKS_CONFIG =
               "${(pkgs.formats.yaml { }).generate "pre-commit-config.yaml" modular.pre-commit.settings.rawConfig}";
           };
 
+          mesonFlags =
+            map (transformFlag "libutil") pkgs.nix-util.mesonFlags
+            ++ map (transformFlag "libstore") pkgs.nix-store.mesonFlags
+            ++ lib.optionals havePerl (map (transformFlag "perl") pkgs.nix-perl-bindings.mesonFlags)
```
Comment (roberth) [inline flake.nix:343]: Thoughts on `transformFlag`:
I think `mesonFlags` should be an implementation detail.
Perhaps this is better implemented as a setup hook, or maybe this is just convenient. Not sure.

File: src/libutil/package.nix [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+
+, pname ? "nix-util"
+
+, versionSuffix ? ""
+, officialRelease ? false
+
+# Check test coverage of Nix. Probably want to use with with at least
+# one of `doCHeck` or `doInstallCheck` enabled.
```
Comment (roberth) [inline src/libutil/package.nix:31]:
```suggestion
# one of `doCheck` or `doInstallCheck` enabled.
```

File: src/libstore/package.nix (hunk 10 above)
Comment (Ericson2314) [reply src/libstore/package.nix:26]: Oh yes, it is much simpler now.

File: flake.nix [hunk 20]
```diff
@@ -236,11 +253,20 @@
...
 
-      packages = forAllSystems (system: rec {
-        inherit (nixpkgsFor.${system}.native) nix changelog-d;
+      packages = forAllSystems (system: let
+        forAllPackagesList = lib.flip map [
+          "nix"
+          "nix-util"
+          "nix-store"
+        ];
+      in rec {
```
Comment (roberth) [inline flake.nix:262]:
```suggestion
      packages = forAllSystems (system: rec {
```

Unused

File: flake.nix [hunk 21]
```diff
@@ -160,21 +160,34 @@
...
-              versionSuffix =
-                if officialRelease
-                then ""
-                else "pre${builtins.substring 0 8 (self.lastModifiedDate or self.lastModified or "19700101")}_${self.shortRev or "dirty"}";
+          nix-util = final.callPackage ./src/libutil/package.nix {
+            inherit
+              fileset
+              stdenv
+              officialRelease
+              versionSuffix
```
Comment (roberth) [inline flake.nix:168]: The version suffix is going to cause unnecessary rebuilds for each commit. I think `officialRelease` is enough "entropy" for the library builds.

File: flake.nix (hunk 18 above)
Comment (Ericson2314) [reply flake.nix:340]: I think eventually in Nixpkgs we should have structured Meson flags, and a nice `shellForPackages` thing that can combine a bunch of things with subprojects and prefix the meson flags of each package.

File: src/libstore/package.nix (hunk 13 above)
Comment (Ericson2314) [reply src/libstore/package.nix:67]: Perhaps, but at the same time, if there is much that needs to be duplicated between the packages, I consider that a Nixpkgs UX problem. E.g. `propagatedBuildInputs` should mean less need to duplicate deps.

File: src/libstore/meson.build (hunk 8 above)
Comment (Ericson2314) [reply src/libstore/meson.build:71]: This specific one is configure-style because one of the BSDs also supports this same interface.

File: src/libstore/meson.build (hunk 7 above)
Comment (Ericson2314) [reply src/libstore/meson.build:171]: Really long term, I hope Meson has a Nix backend, and the Nix backend (+ dynamic derivations) can help Meson support this without these trade-offs.

**Comment by Ericson2314** [comment]
@edolstra 

> I'm not saying I'm against switching to Meson, but I'm not convinced that the approach in this PR is the way to go. 

> and unlike the makefile it's very imperative

Yes, this is my least favorite thing about Meson, but I think it's worth all the benefits.

> It adds a significant amount of complexity to our build system: src/libstore/meson.build is 446 lines (compared to 104 lines for src/libstore/local.mk), src/libstore/package.nix is 158 lines, src/libutil/meson.build is 287 lines, etc. It doesn't make me feel confident that switching to Meson will make Nix easier to maintain.

Comparing `meson.build` vs `local.mk` is not really correct, though, because `meson.build` contains `configure.ac` stuff too. When everything including removing `mk` is factored in, I think the line count diff will be quite positive.

> There is probably a use case for having libstore exposed as a separate derivation (though for most users, having a lib output should be enough), but libutil is strictly internal to the Nix project and is not intended as an external artifact.

I am willing to reconsider these things, but (somewhat ironically) building things together with this approach would be *more* churn, since we would need to have e.g. `src/libstore/libutil` and `src/libstore/libstore`. I think the best approach is to err on the side of two many packages, and then reconsider what the boundaries should be after the old build system is removed.

**Comment by Ericson2314** [comment]
We think the problem is the SQL preprocessing in libstore; I have reverted this PR to just libutil for now, and put the rest in `meson-libstore`.

File: meson.build (hunk 17 above)
[the remark this answers is not in the document]
Comment (Ericson2314) [reply meson.build:6]: I think I rather document when we are done and removing the old build system, because having both build systems documented will be unwieldy.

File: src/libutil/package.nix [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+, libsodium
+, nlohmann_json
+, openssl
+
+# Configuration Options
+
+, versionSuffix ? ""
+, officialRelease ? false
+
+# Check test coverage of Nix. Probably want to use with with at least
```
Comment (roberth) [inline src/libutil/package.nix:23]:
```suggestion
# Check test coverage of Nix. Probably want to use with at least
```

File: src/libutil/package.nix [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+
+  propagatedBuildInputs = [
+    libarchive
+    nlohmann_json
+  ];
+
+  disallowedReferences = [ boost ];
+
+  preConfigure =
+    # "Inline" .version so its not a symlink, and includes the suffix
```
Comment (roberth) [inline src/libutil/package.nix:82]:
```suggestion
    # "Inline" .version so it's not a symlink, and includes the suffix
```

**Review by roberth — APPROVED** [review]

File: flake.nix [hunk 24]
```diff
@@ -236,11 +238,29 @@
...
+        // lib.concatMapAttrs
+          (stdenvName: {}: {
+            "${pkgName}-${stdenvName}" = nixpkgsFor.${system}.stdenvs."${stdenvName}Packages".${pkgName};
+          })
+          (lib.genAttrs stdenvs (_: { })))
+        {
+          "nix" = { };
+          "nix-util" = { };
+        }
+        // lib.optionalAttrs (builtins.elem system linux64BitSystems) {
```
Comment (drupol) [inline flake.nix:263]: Why is this restricted to `linux64BitSystems`?
Also, wouldn't it be a better option to set `meta.platforms` inside the `docker.nix` file ?

Comment (Ericson2314) [reply flake.nix:263]: I think I was just imitating how it worked before?
